sdk/python/feast/go_server.py

```diff
@@ -0,0 +1,246 @@
+        except grpc.RpcError as rpc_error:
+            # If the server became unavailable, it could mean that the subprocess died
+            # or fell into a bad state, so the resolution is to establish a new Go
+            # subprocess and set up a new connection with it.
+
+            # Socket might not have closed if this is a grpc problem.
+            self.stop()
+            if rpc_error.code() == grpc.StatusCode.UNAVAILABLE:
+                self._connect()
+                # Retry request with the new Go subprocess
+                response = self.client.GetOnlineFeatures(request=request)
+            else:
+                error_message = rpc_error.details()
+                if error_message.startswith(FeatureNameCollisionError.__name__):
+                    parsed_error_message = error_message.split(": ")[1].split("; ")
+                    collided_feature_refs = parsed_error_message[0].split(", ")
+                    full_feature_names = parsed_error_message[1] == "true"
+                    raise FeatureNameCollisionError(
+                        collided_feature_refs, full_feature_names
+                    )
+                elif error_message.startswith(ValueError.__name__):
+                    parsed_error_message = error_message.split(": ")[1]
+                    raise ValueError(parsed_error_message)
+                else:
+                    raise
+
+        return OnlineResponse(response)
+
+    def _connect(self):
+        """Start the Go subprocess on a random unused port and connect to it."""
+        # We pass a random unused port to the Go subprocess, so there are no conflicts
+        # if multiple Python processes start a Go subprocess on the same host
+        self.unused_port = _get_unused_port()
+        env = {
+            "FEAST_REPO_CONFIG": self._config.json(),
+            "FEAST_REPO_PATH": self._repo_path,
+            "FEAST_GRPC_PORT": self.unused_port,
+            **os.environ,
+        }
+        cwd = feast.__path__[0]
+
+        if "dev" in feast.__version__:
+            self.process = subprocess.Popen(
+                ["go", "run", "github.com/feast-dev/feast/go/server"],
+                cwd=cwd,
+                env=env,
+                stdin=subprocess.PIPE,
+            )
+        else:
+            goos = platform.system().lower()
+            goarch = "amd64" if platform.machine() == "x86_64" else "arm64"
+            executable = feast.__path__[0] + f"/binaries/go_server_{goos}_{goarch}"
+            self.process = subprocess.Popen(
+                [executable], cwd=cwd, env=env, stdin=subprocess.PIPE
+            )
+
+        # Make sure the subprocess is terminated when the parent process dies
+        # Note: this doesn't handle cases where the parent process is abruptly
+        # killed (e.g. with SIGKILL)
+        atexit.register(lambda: self.stop())
+        self.start_grpc_server()
+        self.pipe_closed = False
+
+    def start_grpc_server(self):
```

Comment:
Is there a reason why `_connect()` is private but `start_grpc_server()` is public?